contracts/src/batch-bridge/L1BatchBridgeGateway.sol

```diff
@@ -0,0 +1,425 @@
+        if (
+            newConfig.maxTxsPerBatch == 0 ||
+            newConfig.maxDelayPerBatch == 0 ||
+            newConfig.feeAmountPerTx > newConfig.minAmountPerTx
+        ) {
+            revert ErrorInvalidBatchConfig();
+        }
+        configs[token] = newConfig;
+    }
+
+    /// @notice Initiate the batch bridge of current pending batch.
+    /// @param token The address of the token.
+    function batchDeposit(address token) external payable onlyRole(KEEPER_ROLE) {
+        BatchConfig memory cachedBatchConfig = configs[token];
+        TokenState memory cachedTokenState = tokens[token];
+        _tryFinalzeCurrentBatch(token, cachedBatchConfig, cachedTokenState);
+
+        // no batch to bridge
+        if (cachedTokenState.currentBatchIndex == cachedTokenState.pendingBatchIndex) {
+            revert ErrorNoPendingBatch();
+        }
+
+        // check bridge fee
+        uint256 depositFee = IL1MessageQueue(queue).estimateCrossDomainMessageFee(cachedBatchConfig.safeBridgeGasLimit);
+        uint256 batchBridgeFee = IL1MessageQueue(queue).estimateCrossDomainMessageFee(SAFE_BATCH_BRIDGE_GAS_LIMIT);
+        if (msg.value < depositFee + batchBridgeFee) {
+            revert ErrorInsufficientMsgValueForBatchDepositFee();
+        }
+
+        // take accumulated fee to fee vault
+        uint256 accumulatedFee;
+        if (token == address(0)) {
+            // no uncheck here just in case
+            accumulatedFee = address(this).balance - msg.value - cachedTokenState.pending;
+        } else {
+            // no uncheck here just in case
+            accumulatedFee = IERC20Upgradeable(token).balanceOf(address(this)) - cachedTokenState.pending;
+        }
+        if (accumulatedFee > 0) {
+            _transferToken(token, feeVault, accumulatedFee);
+        }
+
+        // deposit token to L2
+        BatchState memory cachedBatchState = batches[token][cachedTokenState.pendingBatchIndex];
+        address l2Token;
+        if (token == address(0)) {
+            IL1ScrollMessenger(messenger).sendMessage{value: cachedBatchState.amount + depositFee}(
+                counterpart,
+                cachedBatchState.amount,
+                new bytes(0),
+                cachedBatchConfig.safeBridgeGasLimit
+            );
+        } else {
+            address gateway = IL1GatewayRouter(router).getERC20Gateway(token);
+            l2Token = IL1ERC20Gateway(gateway).getL2ERC20Address(token);
+            IERC20Upgradeable(token).safeApprove(gateway, 0);
+            IERC20Upgradeable(token).safeApprove(gateway, cachedBatchState.amount);
+            IL1ERC20Gateway(gateway).depositERC20{value: depositFee}(
+                token,
+                counterpart,
+                cachedBatchState.amount,
+                cachedBatchConfig.safeBridgeGasLimit
+            );
+        }
+
+        // notify `L2BatchBridgeGateway`
+        IL1ScrollMessenger(messenger).sendMessage{value: batchBridgeFee}(
```

Comment:
What's the rationale for having a 2nd call to the bridge (as opposed to adding this calldata to the 1st call)? Is it that this way we can use a single `finalizeBatchDeposit` function on L2 for both ETH and ERC20 tokens?

Reply:
also noticed this, if `one of them is skipped` but `the other one is successfully relayed`, would this happen?

2 cases:

if `fund transfer` is skipped and `message relayed` is successful, can trigger refund to L1BatchBridge, may need to upgrade the contract, supporting refund to each deposit users in this batch.

if `message relayed` is skipped and `fund transfer` is skipped, how to solve this?

Reply:
A general assumption is that both messages won't be skipped. If so, I think our whole layer2 would have a big bug and all other bridge messages will also be skipped.

Reply:
oh, got it.

Reply:
Why do we need `depositERC20AndCall`? I thought how it works is:
1. Tokens are bridged to L2 receiver aka `L2BatchBridgeGateway` (https://github.com/scroll-tech/scroll/blob/develop/contracts/src/L2/gateways/L2CustomERC20Gateway.sol#L100)
2. The `ERC20Gateway` calls `L2BatchBridgeGateway.onScrollGatewayCallback`.

While the scenario that Colin mentioned is unlikely to happen, it's very concerning: "fund transfer is skipped and message relayed successful" would lead to inconsistency on L1 and L2, and it would be hard to recover from. This seems worse than "other bridge messages will also be skipped".

Reply:
like said before, not all ERC20 gateway will call `L2BatchBridgeGateway.onScrollGatewayCallback`. The skip feature is designed to skip transactions that cannot be proved. In our case, it is very unlikely to happen.

Let's say, if it happens. We can upgrade the contract and do manually refund.

Reply:
also, if we worry about this. We should also worry that `L1BatchBridgeGateway.executeBatchDeposit` and `L2BatchBridgeGateway.distribute` can only be called by permissioned keeper. If we stop keeper program, funds will be locked in the contract.

Reply:
It makes sense.

Reply:
Actually why don't we make `L1BatchBridgeGateway.executeBatchDeposit` and `L2BatchBridgeGateway.distribute` permissionless?

Reply:
permissionless is good but requires a different design (i.e. how to incentive callers) and more codes. If we go permissionless, we should definitely consider about how to handle skipped messages and the potential security problems.

It is a new feature, we should be more careful. In the future, we may change from permissioned to permissionless.